### ISSUE TITLE
`define-macros-order`: option to allow compile only TypeScript code between macros

### ISSUE THREAD

**Issue by andreww2012** [vuejs/eslint-plugin-vue#2978]
**What rule do you want to change?**

`define-macros-order`

**Does this change cause the rule to produce more or fewer warnings?**

Less

**How will the change be implemented? (New option, new default behavior, etc.)?**

Sometimes it's useful to put some types between macro definitions (for example, if they're only used inside the macro following the type declaration):

```vue
<script setup lang="ts">
defineProps(/* ... */)

type Foo = 'bar';

defineEmits(/* ... */)
</script>
```

However, currently any type declaration between macros are disallowed by this rule. It would be nice to introduce an option, say `allowTypesBetweenMacros`, to control whether non-runtime TS code will be allowed between macros.

**Please provide some example code that this change will affect:**

See above

**What does the rule currently do for this code?**

Reports

**What will the rule do after it's changed?**

Will not report

**Additional context**

**Comment by FloEdelmann** [comment]
PR welcome 🙂

**Comment by ValentinYoushkevich** [comment]
Opened #3120 with `allowTypesBetweenMacros`, off by default. Enums and non-ambient namespaces stay disallowed there since they emit runtime code — there's an open question in the PR about whether the existing `skipStatements` list should be aligned with that.
